packages/opal-common/opal_common/authentication/authenticator.py

```diff
@@ -0,0 +1,62 @@
+from abc import abstractmethod
+from typing import Optional
+
+from fastapi import Header
+from opal_common.authentication.deps import JWTAuthenticator
+from opal_common.authentication.types import JWTClaims
+from opal_common.authentication.verifier import JWTVerifier, Unauthorized
+from opal_common.config import opal_common_config
+from opal_common.logger import logger
+
+from .oauth2 import CachedOAuth2Authenticator, OAuth2ClientCredentialsAuthenticator
+
+
+class Authenticator:
+    @property
+    def enabled(self):
+        return self._delegate().enabled
+
+    async def authenticate(self, headers):
+        if hasattr(self._delegate(), "authenticate") and callable(
```

Comment:
That's not a great way to achieve polymorphism. 
What you probably want is to have an Authenticator interface/abstract class.
Have `authenticate` & `enabled` simply raise `NotImplementedError()` (or use `@abc.abstractmethod`, although we don't really use it in OPAL's code base).

Than have different classes inherit `Authenticator` and have those directly implement those methods. (No need to access them through delegation.

Reply:
I've refactored Authenticator as an interface.